File: src/jray/common/Plane.java
package jray.common;

import jray.math.Constants;
import jray.math.VectMatrix;
import jray.math.intersections.RayPlane;

public class Plane extends Object3D {

    public Plane(Vect3 position, Vect3 normal, int color, double reflectivity) {
        this(position, normal);
        this.color = color;
        this.reflectivity = reflectivity;
    }

    public Plane(Vect3 position, Vect3 normal, int color) {
        this(position, normal);
        this.color = color;
    }

    public Plane(Vect3 position, Vect3 normal) {
        super(position, normal);
        normal.normalize();
    }

    @Override
    public double getHitPointDistance(Ray r) {
        double ret = RayPlane.getHitPointRayPlaneDistance(r.getOrigin(), r.getDirection(), position, lookAt);
        if (ret <= Constants.MIN_DISTANCE) {
            return Double.POSITIVE_INFINITY;
        }
        return ret;
    }

    @Override
    public void getNormalAt(Vect3 hitPoint, Vect3 normal) {
        lookAt.copyDataTo(normal);
    }

    @Override
    public boolean contains(Vect3 hitPoint) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public void rotate(Matrix4 tmp) {
        VectMatrix.multiply(lookAt, tmp, lookAt);
    }

    @Override
	public Sphere getBoundingSphere() {
		return null;
	}

	@Override
	public double getBoundingSphereRadius() {
		return Double.POSITIVE_INFINITY;
	}
}

File: JRayXLib/JRayXLib/Shapes/Plane.cs
using System;
using JRayXLib.Colors;
using JRayXLib.Math;
using JRayXLib.Math.intersections;

namespace JRayXLib.Shapes
{
    public class Plane : Basic3DObject {

        public Plane(Vect3 position, Vect3 normal, Color color, double reflectivity)
            : this(position, normal)
        {
            Color = color;
            Reflectivity = reflectivity;
        }

        public Plane(Vect3 position, Vect3 normal, Color color)
            : this(position, normal)
        {
            Color = color;
        }

        public Plane(Vect3 position, Vect3 normal) : base(position, normal){
            LookAt = LookAt.Normalize();
        }

        public override double GetHitPointDistance(Ray r) {
            double ret = RayPlane.GetHitPointRayPlaneDistance(r.GetOrigin(), r.Direction, Position, LookAt);
            if (ret <= Constants.MinDistance) {
                return double.PositiveInfinity;
            }
            return ret;
        }

        public override Vect3 GetNormalAt(Vect3 hitPoint)
        {
            return new Vect3(LookAt);
        }

        public override bool Contains(Vect3 hitPoint) {
            throw new Exception("Not supported yet.");
        }

        public override void Rotate(Matrix4 tmp) {
            VectMatrix.Multiply(LookAt, tmp, ref _lookAt);
        }

        public new Sphere GetBoundingSphere() {
            return null;
        }


        public override double GetBoundingSphereRadius()
        {
            return double.PositiveInfinity;
        }
    }
}

